File: voice_loop/console/panels/schedule.py
"""面板：日程 / 闹钟（同一张表，本项目里它们本来就是一种东西）。

★这里的「加一条」跟语音里说的走**同一套解析器**★（``voice_loop/event_text.py``）：
所以网页里打「明天下午三点组会，提前半小时」和对着麦克风说这句，落地的事件完全一样，
文案也是同一句（``render_added``）。不自己重写一份解析，就不会出现「网页能懂、语音不懂」。

写入一律走 :class:`~voice_loop.events.EventStore`（拿它的锁与 id 规则），
不直接写 JSON 文件——服务那边的提醒线程也在改这个文件（见 store.py 顶部说明）。
"""

from __future__ import annotations

from datetime import datetime, timedelta

from ..registry import Panel

PANEL = Panel(
    id="schedule",
    title="日程 / 闹钟",
    order=20,
    hint="查看、增删改；一句话添加；周视图",
)

CATEGORIES = [
    {"value": "", "label": "（无标签）"},
    {"value": "course", "label": "课程"},
    {"value": "meeting", "label": "会议"},
    {"value": "task", "label": "任务"},
    {"value": "activity", "label": "活动"},
]

REPEATS = [
    {"value": "once", "label": "只一次（闹钟）"},
    {"value": "daily", "label": "每天"},
    {"value": "weekdays", "label": "每个工作日"},
    {"value": "weekly", "label": "每周"},
    {"value": "biweekly", "label": "每两周"},
    {"value": "monthly", "label": "每月"},
    {"value": "yearly", "label": "每年"},
    {"value": "interval", "label": "按间隔"},
]

WEEKDAYS = [
    {"value": i, "label": f"周{'一二三四五六日'[i]}"} for i in range(7)
]


def register(app, ctx) -> None:
    from fastapi import Body, HTTPException  # noqa: PLC0415

    from ... import event_text, events  # noqa: PLC0415

    # ------------------------------------------------------------------ 读
    @app.get("/api/events/schema")
    def api_schema():
        """表单用的选项（前端不写死，以后加周期类型只改这里）。"""
        return {"repeats": REPEATS, "categories": CATEGORIES, "weekdays": WEEKDAYS}

    @app.get("/api/events")
    def api_list(q: str = "", category: str = "", kind: str = "all"):
        """全部事件 + 计算出来的「下一次什么时候、还剩多久」。"""
        store = ctx.events()
        now = datetime.now()
        out = []
        for item in store.load():
            occ = events.first_after(item, now - timedelta(seconds=1))
            text = events.describe(item)
            if q and q not in text:
                continue
            if category and str(item.get("category") or "") != category:
                continue
            repeating = events.has_repeat(item)
            if kind == "alarm" and repeating:
                continue
            if kind == "repeat" and not repeating:
                continue
            out.append({
                "id": item.get("id"),
                "title": events.display_title(item),
                "raw_title": events.title_of(item),
                "category": item.get("category") or "",
                "start": item.get("start") or "",
                "time": item.get("time") or "",
                "repeat": events.repeat_of(item),
                "repeat_text": events.repeat_text(item),
                "weekday": item.get("weekday"),
                "leads": events.leads_of(item),
                "duration_minutes": int(item.get("duration_minutes") or 0),
                "location": item.get("location") or "",
                "note": item.get("note") or "",
                "until": item.get("until") or "",
                "chain": item.get("chain") or None,
                "describe": text,
                "repeating": repeating,
                "needs_confirm": events.needs_confirm(item),
                "next_at": occ.strftime("%Y-%m-%d %H:%M") if occ else "",
                "in_minutes": int((occ - now).total_seconds() // 60) if occ else None,
                "skipped": events.skipped_of(item),
                "state": events.state_of(item),
                "item": item,          # 原样给前端（编辑表单要回填）
            })
        out.sort(key=lambda x: (x["next_at"] or "9999", str(x["id"])))
        return {
            "now": now.strftime("%Y-%m-%d %H:%M:%S"),
            "items": out,
            # ★手工编辑过 events.json 会留下没有 id 的条目★：那种条目改不了也删不了
            # （按 id 找），界面会提示点「整理数据」把它们补上 id。
            "missing_id": sum(1 for row in out if row["id"] is None),
        }

    @app.post("/api/events/tidy")
    def api_tidy():
        """给缺 id 的条目补上 id（就是拿 EventStore.save 过一遍）。

        ``EventStore.save`` 会给缺 id 的条目发新 id（见 events.py 里的注释：
        实测跳过这一步时，一批没 id 的条目会被当成同一个 None，一条指令就把整片删了）。
        """
        store = ctx.events()
        with store.locked():
            items = store.load(force=True)
            before = sum(1 for it in items if it.get("id") is None)
            store.save(items)                      # normalize + 补 id
            after = sum(1 for it in store.load(force=True) if it.get("id") is None)
        return {"ok": True, "fixed": before - after, "before": before}

    @app.get("/api/events/week")
    def api_week(start: str = "", days: int = 7):
        """周视图：从 start（默认今天）起 days 天，每天有哪些发生。"""
        store = ctx.events()
        try:
            base = datetime.strptime(start, "%Y-%m-%d").date() if start else datetime.now().date()
        except ValueError:
            raise HTTPException(status_code=400, detail="start 要写成 YYYY-MM-DD") from None
        days = max(1, min(31, int(days)))
        window_start = datetime.combine(base, datetime.min.time())
        window_end = window_start + timedelta(days=days)
        buckets = {i: [] for i in range(days)}
        for item in store.load():
            for occ in events.occurrences(item, window_start - timedelta(days=1), limit=40):
                if not (window_start <= occ < window_end):
                    continue
                buckets[(occ.date() - base).days].append({
                    "id": item.get("id"),
                    "title": events.display_title(item),
                    "time": occ.strftime("%H:%M"),
                    "at": occ.strftime("%Y-%m-%d %H:%M"),
                    "category": item.get("category") or "",
                    "duration_minutes": int(item.get("duration_minutes") or 0),
                    "done": events.is_done(item, occ),
                })
        for v in buckets.values():
            v.sort(key=lambda x: x["time"])
        today = datetime.now().date()
        return {
            "base": base.isoformat(),
            "days": [
                {
                    "date": (base + timedelta(days=i)).isoformat(),
                    "weekday": (base + timedelta(days=i)).strftime("%Y-%m-%d"),
                    "label": "周" + "一二三四五六日"[(base + timedelta(days=i)).weekday()],
                    "is_today": (base + timedelta(days=i)) == today,
                    "items": buckets[i],
                }
                for i in range(days)
            ],
        }

    # ------------------------------------------------------------------ 增
    @app.post("/api/events")
    def api_add(payload: dict = Body(default={})):
        """两种加法：``{"text": "明天下午三点组会，提前半小时"}`` 或 ``{"fields": {...}}``。"""
        body = payload or {}
        now = datetime.now()
        text = str(body.get("text") or "").strip()
        if text:
            fields = event_text.extract(text, now)
            item = event_text.to_item(fields)
            if not item.get("title") and not body.get("allow_empty_title"):
                # 只说「定个闹钟」也不是错——但网页里得让用户知道标题是空的
                item["title"] = ""
        elif isinstance(body.get("fields"), dict):
            item = _item_from_fields(body["fields"])
        else:
            raise HTTPException(status_code=400, detail="要么给 text（一句话），要么给 fields（表单）")

        store = ctx.events()
        with store.locked():                      # ★读-改-写整段持锁★（服务也在写这个文件）
            saved = store.append(item)
        return {"ok": True, "item": saved, "sentence": _sentence(events, event_text, saved, now)}

    # ------------------------------------------------------------------ 改
    @app.patch("/api/events/{eid}")
    def api_update(eid: str, payload: dict = Body(default={})):
        store = ctx.events()
        old = store.by_id(eid)
        if old is None:
            raise HTTPException(status_code=404, detail=f"没有 id={eid} 的事件")
        body = dict(payload or {})
        confirm = bool(body.pop("confirm", False))
        if events.needs_confirm(old) and not confirm:
            raise HTTPException(
                status_code=409,
                detail="这是一条重复事件，改动会影响往后每一次，需要确认（confirm=1）",
            )
        fields = body.get("fields") if isinstance(body.get("fields"), dict) else body
        patch = _patch_from_fields(fields)
        if not patch:
            raise HTTPException(status_code=400, detail="没有可改的字段")
        updated = store.update_by_id(eid, **patch)
        if updated is None:
            raise HTTPException(status_code=404, detail=f"没有 id={eid} 的事件")
        now = datetime.now()
        return {"ok": True, "item": updated, "sentence": _sentence(events, event_text, updated, now, changed=True)}

    @app.delete("/api/events/{eid}")
    def api_delete(eid: str, confirm: bool = False):
        store = ctx.events()
        item = store.by_id(eid)
        if item is None:
            raise HTTPException(status_code=404, detail=f"没有 id={eid} 的事件")
        if events.needs_confirm(item) and not confirm:
            raise HTTPException(
                status_code=409,
                detail=f"「{events.display_title(item)}」是重复事件（{events.repeat_text(item)}），"
                       "删掉会影响往后每一次，需要确认（confirm=1）",
            )
        with store.locked():
            removed = store.remove_where(lambda it: str(it.get("id")) == str(eid))
        return {"ok": bool(removed), "removed": removed[0] if removed else None}

    # ------------------------------------------------------------------ 单次动作
    @app.post("/api/events/{eid}/skip")
    def api_skip(eid: str, payload: dict = Body(default={})):
        """跳过**某一次**（不是删事件）：下周三不去上课那种。"""
        return _single_shot(ctx, events, eid, payload, action="skip")

    @app.post("/api/events/{eid}/done")
    def api_done(eid: str, payload: dict = Body(default={})):
        """标记某一次已完成（事件链靠它：「A 完成后再提醒 B」）。"""
        return _single_shot(ctx, events, eid, payload, action="done")

    @app.post("/api/events/{eid}/undone")
    def api_undone(eid: str, payload: dict = Body(default={})):
        return _single_shot(ctx, events, eid, payload, action="undone")


# ---------------------------------------------------------------- 内部工具


def _single_shot(ctx, events, eid: str, payload: dict, *, action: str) -> dict:
    """skip / done / undone 共用的实现（都是「改某一个发生时刻的状态」）。"""
    from fastapi import HTTPException  # noqa: PLC0415

    store = ctx.events()
    item = store.by_id(eid)
    if item is None:
        raise HTTPException(status_code=404, detail=f"没有 id={eid} 的事件")
    now = datetime.now()
    at = events.parse_dt((payload or {}).get("at")) or events.first_after(
        item, now - timedelta(minutes=5)
    ) or events.parse_dt(item.get("start"))
    if at is None:
        raise HTTPException(status_code=400, detail="这条事件没有可操作的时间点")

    state = dict(item.get("state") or {})
    key = at.strftime("%Y-%m-%d %H:%M:%S")
    day_key = at.strftime("%Y-%m-%d")

    if action == "skip":
        days = [d for d in (state.get("skipped") or []) if str(d) != day_key]
        days.append(day_key)
        state["skipped"] = days[-40:]
        text = f"已经跳过 {day_key} 这一次"
    elif action == "done":
        done = [d for d in (state.get("done") or []) if str(d) != key]
        done.append(key)
        state["done"] = done[-40:]
        text = f"已标记 {key} 完成"
    else:
        state["done"] = [d for d in (state.get("done") or []) if str(d) != key]
        text = f"已取消 {key} 的完成标记"

    store.update_by_id(eid, state=state)
    return {"ok": True, "at": key, "message": text, "state": state}


def _item_from_fields(fields: dict) -> dict:
    """表单字段 → 事件条目（走 new_event，保持文件干净：空值不写进去）。"""
    from ... import events  # noqa: PLC0415

    start = str(fields.get("start") or "").strip()
    if start and len(start) == 16:              # 前端 datetime-local 给的是 2026-09-25T15:30
        start = start.replace("T", " ") + ":00"
    repeat = str(fields.get("repeat") or "once").strip() or "once"
    leads = fields.get("remind_before")
    if isinstance(leads, str):
        leads = [int(x) for x in leads.replace("，", ",").split(",") if x.strip().lstrip("-").isdigit()]
    return events.new_event(
        str(fields.get("title") or "").strip(),
        start=start,
        repeat=repeat,
        remind_before=[int(x) for x in (leads or [0])],
        duration_minutes=int(fields.get("duration_minutes") or 0),
        location=str(fields.get("location") or "").strip(),
        note=str(fields.get("note") or "").strip(),
        category=str(fields.get("category") or "").strip(),
        until=str(fields.get("until") or "").strip(),
        weekday=_int_or_none(fields.get("weekday")),
        month=_int_or_none(fields.get("month")),
        day=_int_or_none(fields.get("day")),
        time=str(fields.get("time") or "").strip(),
    )


def _patch_from_fields(fields: dict) -> dict:
    """编辑时的改动集。**空串会被 EventStore.normalize 当成「删掉这个字段」**。"""
    out: dict = {}
    for key in ("title", "category", "location", "note", "until", "time", "repeat", "start"):
        if key in fields:
            value = fields[key]
            if key == "start" and isinstance(value, str) and len(value) == 16:
                value = value.replace("T", " ") + ":00"
            out[key] = "" if value is None else str(value)
    for key in ("duration_minutes",):
        if key in fields:
            out[key] = int(fields[key] or 0)
    if "weekday" in fields:
        out["weekday"] = _int_or_none(fields["weekday"])
    if "remind_before" in fields:
        leads = fields["remind_before"]
        if isinstance(leads, str):
            leads = [x.strip() for x in leads.replace("，", ",").split(",")]
        out["remind_before"] = [int(x) for x in (leads or []) if str(x).lstrip("-").isdigit()]
    return out


def _int_or_none(value):
    try:
        return int(value)
    except (TypeError, ValueError):
        return None


def _sentence(events, event_text, item: dict, now: datetime, *, changed: bool = False) -> str:
    """给用户看的一句话（★跟语音回复用同一个 render★，两边说法一致）。"""
    start = events.first_after(item, now - timedelta(minutes=5)) or events.parse_dt(item.get("start"))
    if start is None:
        return events.describe(item)
    try:
        return event_text.render_added(item, start, now, changed=changed)
    except Exception:  # noqa: BLE001 - 文案生成失败不影响数据
        return events.describe(item)

File: voice_loop/console/static/panels/schedule.js
/* 日程 / 闹钟面板：列表 + 一句话添加 + 编辑表单 + 周视图。 */
(function () {
  let handleEvent = null;

  window.Console.register("schedule", {
    render(root, ctx) {
      const { h, card, table, field, collect } = UI;
      let data = { items: [] };
      let schema = { repeats: [], categories: [], weekdays: [] };
      let weekStart = null;      // 周视图基准日期（null = 今天）
      let editing = null;        // 正在编辑的事件 id

      // ---------------- 一句话添加 ----------------
      const quick = h("input", { placeholder: "比如：明天下午三点组会，提前半小时 / 每周三 9 点上课" });
      const quickAdd = async () => {
        const text = quick.value.trim();
        if (!text) return;
        const r = await ctx.api.safe(() => ctx.api.post("/api/events", { text }));
        if (!r) return;
        quick.value = "";
        ctx.toast(r.sentence || "已添加", "ok");
        refresh();
      };
      quick.addEventListener("keydown", (e) => { if (e.key === "Enter") quickAdd(); });

      // ---------------- 结构化表单（新增/编辑共用） ----------------
      const formBox = h("div", {});

      function openForm(item) {
        editing = item ? item.id : null;
        const it = (item && item.item) || {};
        const inputs = {};
        const f = h("form", { class: "grid" });
        const add = (key, label, config) => {
          const wrap = field(label, config, null);
          wrap.dataset.key = key;
          inputs[key] = wrap;
          f.appendChild(wrap);
        };
        add("title", "内容（留空=纯闹钟）", { value: it.title || "" });
        add("start", "时间（一次性事件：日期+时刻）", { type: "datetime-local", value: _toLocalInput(it.start) });
        add("repeat", "重复", {
          type: "select", value: it.repeat || "once",
          options: schema.repeats.map((r) => ({ value: r.value, label: r.label })),
        });
        add("time", "钟点（★重复事件用这个★ HH:MM）", { type: "time", value: it.time || "" });
        add("weekday", "星期（每周/每两周用）", {
          type: "select", value: it.weekday === undefined || it.weekday === null ? "0" : String(it.weekday),
          options: schema.weekdays.map((w) => ({ value: String(w.value), label: w.label })),
        });
        add("remind_before", "提前量（分钟，逗号分隔；0=准时）", {
          value: (it.remind_before || [0]).join(","),
        });
        add("duration_minutes", "时长（分钟，0=不占时间）", { type: "number", value: it.duration_minutes || 0, min: 0 });
        add("category", "标签", {
          type: "select", value: it.category || "",
          options: schema.categories.map((c) => ({ value: c.value, label: c.label })),
        });
        add("location", "地点", { value: it.location || "" });
        add("note", "备注（提醒时会念出来）", { value: it.note || "" });
        add("until", "截止日期（YYYY-MM-DD）", { value: it.until || "" });

        const submit = h("button", {
          class: "btn primary", type: "submit", text: editing ? "保存改动" : "添加",
        });
        const cancel = h("button", {
          class: "btn ghost", type: "button", text: "取消",
          onclick: () => { formBox.innerHTML = ""; editing = null; },
        });

        const doSubmit = async () => {
          const payload = collect(f);
          if (editing) {
            const ok = await ctx.api.safe(
              () => ctx.api.patch("/api/events/" + editing, { fields: payload, confirm: true }),
              "已保存");
            if (ok) { formBox.innerHTML = ""; editing = null; refresh(); }
          } else {
            const r = await ctx.api.safe(() => ctx.api.post("/api/events", { fields: payload }));
            if (r) { ctx.toast(r.sentence || "已添加", "ok"); formBox.innerHTML = ""; refresh(); }
          }
        };
        UI.onSubmit(f, doSubmit);
        f.appendChild(h("div", { class: "row" }, [submit, cancel]));
        formBox.innerHTML = "";
        formBox.appendChild(card(editing ? "编辑事件 #" + editing : "手动添加一条", f));
      }

      // ---------------- 列表 ----------------
      const listBox = h("div", {});
      const weekBox = h("div", {});

      function renderList() {
        listBox.innerHTML = "";
        const items = data.items || [];
        if (data.missing_id) {
          listBox.appendChild(card("需要整理一下数据", [
            h("p", { text: "有 " + data.missing_id + " 条事件没有 id（多半是手工编辑过 data/events.json）。"
              + "没有 id 就没法改/删（那些按钮点了会报错）。点一下右边就能补上——不会动其他字段。" }),
            h("div", { class: "row" }, [
              h("button", {
                class: "btn primary", text: "补 id（整理数据）",
                onclick: async () => {
                  const r = await ctx.api.safe(() => ctx.api.post("/api/events/tidy", {}));
                  if (r) { ctx.toast("已补 " + r.fixed + " 条", "ok"); refresh(); }
                },
              }),
            ]),
          ]));
        }
        listBox.appendChild(card("全部事件（" + items.length + " 条）· 现在 " + (data.now || ""), [
          h("div", { class: "row" }, [
            h("button", {
              class: "btn", text: "只显示重复的",
              onclick: () => { filter.kind = filter.kind === "repeat" ? "all" : "repeat"; refresh(); },
            }),
            h("button", {
              class: "btn", text: "只显示一次性的",
              onclick: () => { filter.kind = filter.kind === "alarm" ? "all" : "alarm"; refresh(); },
            }),
            h("span", { class: "muted", text: "当前筛选：" + (filter.kind === "all" ? "全部" : filter.kind) }),
          ]),
          table([
            { key: "id", label: "#", cls: "num", render: (r) => (r.id === null || r.id === undefined ? "无" : r.id) },
            { key: "next_at", label: "下一次", cls: "num", render: (r) => r.next_at
                ? h("div", {}, [h("div", { text: r.next_at }),
                                r.in_minutes !== null ? h("span", { class: "muted", text: (r.in_minutes >= 0 ? "还有 " : "已过 ") + Math.abs(r.in_minutes) + " 分" }) : null])
                : h("span", { class: "muted", text: "—" }) },
            { key: "title", label: "内容", render: (r) => h("div", {}, [
                h("div", { text: r.title }),
                r.location ? h("span", { class: "muted", text: "📍" + r.location }) : null,
                r.note ? h("span", { class: "muted", text: " · " + r.note }) : null,
              ]) },
            { key: "repeat_text", label: "重复", render: (r) => r.repeat_text || "一次" },
            { key: "leads", label: "提前", render: (r) => r.leads.map(leadText).join(" / ") },
            { key: "category", label: "标签", render: (r) => (r.category ? h("span", { class: "tag", text: r.category }) : "") },
            { key: "state", label: "状态", render: (r) => {
                const done = (r.state.done || []).length, fired = (r.state.fired || []).length;
                const skipped = (r.state.skipped || []).length;
                return h("div", { class: "row" }, [
                  done ? h("span", { class: "tag ok", text: "完成 " + done }) : null,
                  fired ? h("span", { class: "tag", text: "已响 " + fired }) : null,
                  skipped ? h("span", { class: "tag warn", text: "跳过 " + skipped }) : null,
                ]);
              } },
            { key: "act", label: "", cls: "act", render: (r) => h("div", { class: "row" }, [
                r.id === null || r.id === undefined
                  ? h("span", { class: "muted", text: "先补 id 才能操作" })
                  : h("div", { class: "row" }, [
                      h("button", { class: "btn small", text: "跳过本次", title: "只跳过下一次，不删事件",
                        onclick: () => act(r, "skip") }),
                      h("button", { class: "btn small", text: "完成", onclick: () => act(r, "done") }),
                      h("button", { class: "btn small", text: "编辑", onclick: () => openForm(r) }),
                      h("button", { class: "btn small danger", text: "删除",
                        onclick: async () => {
                          const warn = r.repeating
                            ? "「" + r.title + "」是重复事件（" + (r.repeat_text || "") + "），删掉会影响往后每一次。确定？"
                            : "删除「" + r.title + "」？";
                          if (!confirm(warn)) return;
                          const ok = await ctx.api.safe(
                            () => ctx.api.del("/api/events/" + r.id + "?confirm=1"), "已删除");
                          if (ok) refresh();
                        } }),
                    ]),
              ]) },
          ], items, { emptyText: "还没有任何事件。用上面那个输入框一句话加一条试试。" }),
        ]));
      }

      async function act(row, action) {
        const r = await ctx.api.safe(() => ctx.api.post("/api/events/" + row.id + "/" + action, {}));
        if (r) { ctx.toast(r.message || "已处理", "ok"); refresh(); }
      }

      // ---------------- 周视图 ----------------
      function renderWeek() {
        weekBox.innerHTML = "";
        ctx.api.get("/api/events/week" + (weekStart ? "?start=" + weekStart : ""))
          .then((w) => {
            const nav = h("div", { class: "row" }, [
              h("button", { class: "btn small", text: "上一周", onclick: () => shiftWeek(-7) }),
              h("button", { class: "btn small", text: "本周", onclick: () => { weekStart = null; renderWeek(); } }),
              h("button", { class: "btn small", text: "下一周", onclick: () => shiftWeek(7) }),
              h("span", { class: "muted grow", text: w.base + " 起 7 天" }),
            ]);
            const grid = h("div", { class: "week" }, w.days.map((d) => h("div", {
              class: "day" + (d.is_today ? " today" : ""),
            }, [
              h("h4", { text: d.date.slice(5) + " " + d.label + (d.is_today ? " ·今天" : "") }),
              ...d.items.map((it) => h("div", { class: "slot", title: it.at }, [
                h("div", { text: it.title }),
                h("span", { class: "t", text: it.time + (it.duration_minutes ? " · " + it.duration_minutes + "分" : "") }),
                it.done ? h("span", { class: "tag ok", text: "完成" }) : null,
              ])),
              d.items.length === 0 ? h("p", { class: "muted", text: "—" }) : null,
            ])));
            weekBox.appendChild(card("周视图", [nav, grid]));
          })
          .catch((err) => { weekBox.appendChild(card("周视图", UI.line("读取失败：" + err.message))); });
      }

      function shiftWeek(days) {
        const base = weekStart ? new Date(weekStart + "T00:00:00") : new Date();
        base.setDate(base.getDate() + days);
        weekStart = base.toISOString().slice(0, 10);
        renderWeek();
      }

      // ---------------- 组装 ----------------
      const filter = { kind: "all" };
      root.appendChild(card("加一条（跟语音说的是同一套解析）", [
        h("div", { class: "row" }, [
          h("div", { class: "grow" }, quick),
          h("button", { class: "btn primary", text: "添加", onclick: quickAdd }),
          h("button", { class: "btn", text: "用表单填", onclick: () => openForm(null) }),
        ]),
        h("p", { class: "muted", text: "支持的说法和语音完全一致：日期、时间、每周几、提前量、时长、地点、备注。解析结果就是落地的事件。" }),
        formBox,
      ]));
      root.appendChild(listBox);
      root.appendChild(weekBox);

      function refresh() {
        ctx.api.get("/api/events?kind=" + filter.kind)
          .then((d) => { data = d; renderList(); })
          .catch((err) => ctx.toast("读取事件失败：" + err.message, "err"));
      }

      ctx.api.get("/api/events/schema").then((s) => { schema = s; });
      refresh();
      renderWeek();

      handleEvent = (ev) => {
        // 服务那边响了提醒会写 state，前端顺手刷新一下列表（不然「已响」看着是旧的）
        if (ev.kind === "log" && /提醒|reminder|fired/.test(ev.data && ev.data.line ? ev.data.line : "")) {
          clearTimeout(renderList._t);
          renderList._t = setTimeout(refresh, 1500);
        }
      };
    },

    onEvent(ev, ctx) { if (handleEvent) handleEvent(ev, ctx); },
    onLeave() { handleEvent = null; },
  });

  /** ISO 字符串 → datetime-local 输入框要的格式（本地时间，不要 Z） */
  function _toLocalInput(value) {
    if (!value) return "";
    const s = String(value).replace(" ", "T");
    return s.length >= 16 ? s.slice(0, 16) : s;
  }

  /** 提前量（分钟）→ 人话：1440 该说「1天」而不是「1440分」 */
  function leadText(minutes) {
    const m = Number(minutes) || 0;
    if (m === 0) return "准时";
    if (m % 1440 === 0) return m / 1440 + "天";
    if (m % 60 === 0) return m / 60 + "小时";
    return m + "分钟";
  }
})();
